Refactor interface items

osheamonn opened the pull request:
## Proposed changes

Refactors the InterfaceItem infrastructure to make better use of base tags, so that decisions about how quantities are computed on the interface are made when adding the Tag to the DataBox (is this a SimpleTag, a ComputeTag, or does it slice variables from the volume?) rather than at the level of the Tags inside of variables which is done now. 

Most of the significant changes occur in src/Domain/Tags.hpp. 
- InterfaceBase and InterfaceImpl removed. 
- InterfaceComputeItem and Slice added.
- Interface is now a simple item, which can be used as a base tag for the above two items. 

Breaks: 
- Anything that touches the Interface part of the Initialization code, the user must now specify 
whether Interface tags are simple, compute items or slices. Before, using the tag Interface worked for everything. 
- If `should_be_sliced_to_boundary` was being defined in any Tags, it now doesn't do anything. Whether the Tag should be sliced is specified by adding it to the DataBox using `Slice`. 

There are some TODO's scattered around which are questions I have for people, regarding using `tmpl::contains_v<TagsList, Tag>` to determine whether a Tag is in the DataBox, since it now fails for the compute tags or slice tags on the interface. 

The last commit (remove sliced_to...) should probably be squashed into the previous one, but I thought it would make reviewing easier.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

osheamonn committed:
Fix minor typos in Databox

osheamonn committed:
Add stream operator for Mesh

This allows Mesh to work correctly when used inside of CHECK's

osheamonn committed:
Link required Libraries for StrahlKorper and TOV

kidder commented on src/Domain/Tags.hpp:
```diff
@@ -155,50 +155,39 @@ struct BoundaryDirections : db::ComputeTag {
+/// If a SimpleTag is desired on the interface, then this tag can be added to
+/// the DataBox directly. If a ComputeTag which acts on Tags on the interface is
+/// desired, then the tag should be added using `InterfaceComputeItem`. If a
+/// ComputeTag which slices a VariablesTag in the volume to an Interface, then
```
Interface is desired,

kidder commented on tests/Unit/Evolution/DiscontinuousGalerkin/Test_InitializeElement.cpp:
```diff
@@ -144,6 +144,15 @@ bool box_contains(const Box& /*box*/) noexcept {
   return tmpl::list_contains_v<typename Box::tags_list, Tag>;
 }
 
+// TODO: Because now (almost) all Interface tags are retrieved by base tags,
```
want to keep a runtime check

osheamonn commented:
Rebased and squashed. `tag_is_retrievable` is now done using a type trait which is checked at runtime.

osheamonn committed:
Refactor Interface Tags

Add InterfaceComputeItem and Slice, which have Interface as a base tag. This allows how an InterfaceItem is computed to be specified when inserted into the DataBox.

osheamonn commented:
Had to rebase due to IWYU @kidder

nilsdeppe commented:
travis failures are test timeouts. @wthrowe feel free to merge if you're happy with this :)
